### ISSUE TITLE
How To Implement The EVP_aes_256_cbc Using the OpenSSL Library in VS2019?

### ISSUE THREAD

**Issue by Ehsan-Yaghoubi** [openssl/openssl#10903]
Hello, I am trying to implement EVP_aes_256_cbc in c++ using the following link: 
[EVP_Symmetric_Encryption_and_Decryption](https://wiki.openssl.org/index.php/EVP_Symmetric_Encryption_and_Decryption)
As it can bee seen in the above link, the "key" and the "plaintext" have been hardcoded as follows:
`unsigned char *key = (unsigned char *)"01234567890123456789012345678901"; `
`unsigned char *plaintext = (unsigned char *)"The quick brown fox jumps over the lazy dog";`

this works fine, but In my application, "key" and the "plaintext" should not be hardcoded, instead, I have to read each of them from the buffer of a signal of type byte (I am using a simulator(its a collection of codes) that reads data from a file and saves it into a signal).

What I did is the following:
1. I read the 16 bytes of the plaintext from the buffer (using the "bufferGet" function) and I stored the data in a byte array of size 16.
2. I read the 32 bytes of the key from the buffer (using the "bufferGet" function) and again I stored the data in a byte array of size 32.
3. Since the inputs of the "encrypt function" are pointers, I should creat pointers to the arrays (i am not sure if I am doing it correctly)
4. Then I call the encrypt function and it provides me the ciphertext.
5. I save the ciphertext into a signal of type byte.

However, I am not having the correct output. I think the problem is with the 3rd step because in the debugger I see some extra garbage at the end of the pointers. please check the following picture:

![WhatsApp Image 2020-01-20 at 12 13 54](https://user-images.githubusercontent.com/36289828/72735837-858c8980-3b94-11ea-8728-78841a965ae5.jpeg)

I have to mention that I am new in programming and what I did may not be correct. I highly appreciate your help in solving the problem.

Here are my codes: (t_raw is equal to byte)
`// # Reading plaintext and key from the signalst_raw alicePlainIn[16]{};`
`inputSignals[0]->bufferGet((std::byte*)alicePlainIn, 16);`
`t_raw aliceKeys[32]{};`
`inputSignals[1]->bufferGet((std::byte*)aliceKeys, 32);`
`// # Encryption Process`
`ciphertext_len = encrypt((unsigned char*)plaintext, strlen((char*)plaintext), key, iv, ciphertext);`
`// # Saving the Ciphertext into a Signal`
`for (auto cipher_text : ciphertext)`
`{`
`outputSignals[0]->bufferPut((t_raw)cipher_text);`
`}`

Thank you!

**Comment by mattcaswell** [comment]
Most likely the problem is related to treating string data as binary data or vice-versa. The EVP_Encrypt* functions expect keys, IVs, plaintext, ciphertext etc to be *binary* data. In C a string is a sequence of bytes terminated by the "NUL" (0) character. Binary data is not necessarily terminated like this and you typically have to pass around "length" values (or in the case of keys/ivs the lengths are fixed and known in advance).

So for example I see in you code that read 16 bytes of plaintext data into the "alicePlainIn" buffer. You then call strlen on it. If the plaintext is binary data and includes the 0 character in it, or is exactly 16 bytes in length then the strlen fuunction is going to go wrong. strlen assumes the input is string data and scans the buffer until it finds the NUL terminator. Quite likely this is running past the end of your alicePlainIn buffer and giving you an incorrect length.

**Comment by Ehsan-Yaghoubi** [comment]
Thank you so much for your detailed explanation.  Yes, in my application the plaintext may contain every sort of data including binary. I applied your changes by assigning the strlen((char*)plaintext) to 16:
`ciphertext_len = encrypt((unsigned char*)plaintext, 16 , key, iv, ciphertext);`

1. Are these changes correct? 
2. Do I need to make changes for the "key" as well? Because I am not having the correct output yet.
3. Suppose that the signals are as follows, what would be the length of the ciphertext?
alicePlainIn= "this is my text."
aliceKey= "01010101010101010101010101010101"

**Comment by mattcaswell** [comment]
> Are these changes correct?

It is difficult to say for sure from the small piece of code that you have shown us, but it looks ok.

> Do I need to make changes for the "key" as well?

Well, again its difficult to say for sure from the small piece of code you've shown us. However there's nothing in what you've shown us that suggests you're treating the key as anything other than 32-bytes of binary data - so it may be ok. For AES-256 the key is a fixed length of 256 bits (32 bytes), so as long as the full 32 bytes are being populated with the correct data then you should be ok.

> Suppose that the signals are as follows, what would be the length of the ciphertext?

The length of the plaintext is always padded to the next multiple of 16. If its already a multiple of 16 then you need 16 bytes of padding. So plaintext of length 1 would give you 16 bytes of ciphertext. Plaintext of length 15 would also give you 16 bytes of ciphertext. Plaintext of length 16 would give you 32 bytes of ciphertext, and so on. In your case the input string (without any NUL terminator) is exactly 16-bytes of data, therefore we would expect to see 32 bytes of ciphertext.

> Because I am not having the correct output yet.

What are you expecting to see?

A couple of things spring to mind that could still be going wrong.

1) You've said nothing about what IV you are using. If you have got some reference output that you are trying to recreate then you must make sure the IV in use is correct.

2) The key you've shown looks like a string containing a repeated sequence of an ASCII "0" character followed by an ASCII "1" character. I assume you really did mean that, and not a repeated sequence of a 0 byte value followed by a 1 byte value - because of course ASCII "0" character is different to 0 byte value!?

**Comment by Ehsan-Yaghoubi** [comment]
Thank you again for the great explanation.  Unfortunately, the simulator is huge and I couldn't include all the codes. However, I only changed the "key" and the "plaintext" and the other variables (including IV) are the same as the OpenSSL structure in the above link. 

> there's nothing in what you've shown us that suggests you're treating the key as anything other than 32-bytes of binary data :  

Since I am using the EVP interface, many details have been hidden and I am not sure how the key is used inside the OpenSSL functions. That's why I asked if I need to make any changes. I searched through the OpenSSL wiki and other online resources, but I could not find useful materials. In case you know any documentation that explains AES C++ implementation, I would appreciate to inform me. 

> What are you expecting to see?

1. In order to test my output, first I used the hardcoded version of the OpenSSL as follows:

`unsigned char* plaintext = (unsigned char*)"This is my text";`
`unsigned char* key = (unsigned char*)"01110101011101010111110101011101";`
`unsigned char* iv = (unsigned char*)"0123456789012345";`

And this is my output (that according to your explanation should be 16 byte and it is):
öžõYß™D–Ÿƒ%nš·

2. Then I tried to test my own version of codes with the same input data (IV, key, plaintext...)

This is my output when I did not change the strlen:  (the first 16 bytes are correct but it keeps generating garbage continuously). 
öžõYß™D–Ÿƒ%nš·‹X¬þ|Ö÷Ç~^CôwFØZfsiÔ²ý9t2¡FØZfsiÔ²ý9t2¡..............garbage

And this is my output when I applied your changes  and used 16 instead of strlen((char*)plaintext): (only creates garbage)
ù|ÁëÊ­†w)Ò(Ö Ëÿ+J”Ûà'{‘î}ªÆ|ðüªCÙŒ@Ä±Æay›ˆ8Ÿ´ð9eÉ¸¥GZ.<a¦Lª#..............garbage

> The key you've shown looks like a string containing a repeated sequence of an ASCII "0" character followed by an ASCII "1" character. I assume you really did mean that, and not a repeated sequence of a 0 byte value followed by a 1 byte value - because of course ASCII "0" character is different to 0 byte value!?

The key that I am using is a sequence of binary data. there is no pattern in the binary values. I am reading the 32 bytes of the key from a  signal (of type byte) and then I store the key into an array.

Thank you so much for your time and corrections.

**Comment by mattcaswell** [comment]
> And this is my output when I applied your changes and used 16 instead of strlen((char*)plaintext): (only creates garbage)

It looks like you're trying to "print" the ciphertext- which is not going to work well because the encrypted data is binary data and not intended to be printed.
What is the length of the ciphertext that is returned from the EVP_Encrypt*() function call?. So if you've implemented the encryption in the same way as shown in the wiki page I'm interested in the value of `ciphertext_len` in the code below after you've attempted your encryption.
````
    if(1 != EVP_EncryptUpdate(ctx, ciphertext, &len, plaintext, plaintext_len))
        handleErrors();
    ciphertext_len = len;

    /*
     * Finalise the encryption. Further ciphertext bytes may be written at
     * this stage.
     */
    if(1 != EVP_EncryptFinal_ex(ctx, ciphertext + len, &len))
        handleErrors();
    ciphertext_len += len;
````

**Comment by Ehsan-Yaghoubi** [comment]
I updated my previous post and made some corrections in section  _What are you expecting to see?_ and I would appreciate if you could check it again.

> What is the length of the ciphertext that is returned from the EVP_Encrypt*() function call?

After the new update, for both hardcoded version and my own version of codes: `ciphertext_len=16`
I would be grateful if you could help me with how to get rid of extra garbages at the end of my output.
Thank you!

**Comment by mattcaswell** [comment]
So, if the ciphertext length is 16, then that is what you are expecting right? As long as the first 16 bytes of the ciphertext output is what you wanted to see then its working. Anything beyond the first 16 bytes will be garbage - don't attempt to print it or look at it!

**Comment by Ehsan-Yaghoubi** [comment]
Yes, that is exactly what I expect. The garbage was related to my simulator and it is fixed now. Thank you so much for your help and detailed explanations.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input
